In [5]:
import pandas as pd
import numpy as np

In [10]:
csv_file = "C:/Hany/Data Analyst Course/paython/Pandas-Chellenge/HeroesOfPymoli_Resources_purchase_data.csv"
Heroes_of_Pymoli = pd.read_csv(csv_file)
new_Item_ID = Heroes_of_Pymoli.groupby('Item ID').count()
Heroes_of_Pymoli.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [11]:
Heroes_of_Pymoli.describe()

Purchase ID         Age     Item ID       Price
count   780.000000  780.000000  780.000000  780.000000
mean    389.500000   22.714103   91.755128    3.050987
std     225.310896    6.659444   52.697702    1.169549
min       0.000000    7.000000    0.000000    1.000000
25%     194.750000   20.000000   47.750000    1.980000
50%     389.500000   22.000000   92.000000    3.150000
75%     584.250000   25.000000  138.000000    4.080000
max     779.000000   45.000000  183.000000    4.990000

In [12]:
new_Item_ID["Price"].sum()

780

In [344]:
#total number of players

In [13]:
Unique_Items = Heroes_of_Pymoli["SN"].drop_duplicates()
Unique_Items.count()

576

In [7]:
#Number of Unique Items

In [14]:
Unique_Items = Heroes_of_Pymoli["Item Name"].drop_duplicates()
Unique_Items.count()

179

In [9]:
#Average Purchase Price

In [15]:
Heroes_of_Pymoli["Price"].mean()

3.050987179487176

In [11]:
#Total Number of Purchases

In [16]:
Heroes_of_Pymoli[["Purchase ID"]].count()

Purchase ID    780
dtype: int64

In [13]:
#Total Revenue

In [17]:
Heroes_of_Pymoli["Price"].sum()

2379.77

In [15]:
#Percentage and Count of Male Players

In [18]:
# dropping ALL duplicte values 
Heroes_of_Pymoli.drop_duplicates(subset ="SN", keep = False, inplace = True)
  
# displaying data 
Heroes_of_Pymoli.head()

Purchase ID             SN  Age Gender  Item ID  \
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   
6            6      Itheria73   36   Male      169   

                                Item Name  Price  
1                       Frenzied Scimitar   1.56  
2                            Final Critic   4.88  
3                             Blindscythe   3.27  
4                                    Fury   1.44  
6  Interrogator, Blood Blade of the Queen   2.18

In [19]:
gender_percent= Heroes_of_Pymoli.Gender
Counts = gender_percent.value_counts()
Percet100 = gender_percent.value_counts(normalize=True).mul(100).round(2).astype(str) + "%"
pd.DataFrame({"Count": Counts, "Percentage": Percet100})

Count Percentage
Male                     352     85.02%
Female                    55     13.29%
Other / Non-Disclosed      7      1.69%

In [37]:
gender= Heroes_of_Pymoli.groupby('Gender').size()
gender

Gender
Female                    55
Male                     352
Other / Non-Disclosed      7
dtype: int64

In [20]:
Heroes_of_Pymoli_males = Heroes_of_Pymoli.query("Gender == 'Male'").SN.unique()
len(Heroes_of_Pymoli_males)
Heroes_of_Pymoli_females = Heroes_of_Pymoli.query("Gender == 'Female'").SN.unique()
len(Heroes_of_Pymoli_females)
Heroes_of_Pymoli_other = Heroes_of_Pymoli.query("Gender == 'Other / Non-Disclosed'").SN.unique()
len(Heroes_of_Pymoli_other)
Heroes_of_Pymoli_all = Heroes_of_Pymoli.SN.unique()
len(Heroes_of_Pymoli_all)

414

In [21]:
new_heroes = Heroes_of_Pymoli[['SN', 'Gender', 'Price']].copy()
new_gender = Heroes_of_Pymoli["SN"].drop_duplicates()
new_gender.head()

1      Lisovynya38
2       Ithergue48
3    Chamassasya86
4        Iskosia90
6        Itheria73
Name: SN, dtype: object

In [22]:
### Purchasing Analysis (Gender/Male)
male_data = Heroes_of_Pymoli.query('Gender == "Male"')
male_purchase_count = male_data["Purchase ID"].count()
male_Average_Purchase_Price = male_data["Price"].mean()
male_Total_Purchase_Value = male_data["Price"].sum()
male_Total_Purchase_Value = male_data["Price"].sum()
print("Purchase count for males is " + str(male_purchase_count))
print("Average purchase price for males is " + str(male_Average_Purchase_Price))
print("Total purchase valuefor males is " + str(male_Total_Purchase_Value))
male_data.loc[male_data['SN'] == 1, 'Price'].sum()
print("Average Purchase Total for male " + str(male_data["Price"].mean()))

Purchase count for males is 352
Average purchase price for males is 3.021477272727272
Total purchase valuefor males is 1063.5600000000002
Average Purchase Total for male 3.021477272727272


In [23]:
### Purchasing Analysis (Gender/Female)
female_data = Heroes_of_Pymoli.query('Gender == "Female"')
female_purchase_count = female_data["Purchase ID"].count()
female_Average_Purchase_Price = female_data["Price"].mean()
female_Total_Purchase_Value = female_data["Price"].sum()
female_Total_Purchase_Value = female_data["Price"].sum()
print("Purchase count for females is " + str(female_purchase_count))
print("Average purchase price for females is " + str(female_Average_Purchase_Price))
print("Total purchase valuefor females is " + str(female_Total_Purchase_Value))
female_data.loc[female_data['SN'] == 1, 'Price'].sum()
print("Average Purchase Total for female " + str(female_data["Price"].mean()))

Purchase count for females is 55
Average purchase price for females is 3.2061818181818182
Total purchase valuefor females is 176.33999999999997
Average Purchase Total for female 3.2061818181818182


In [24]:
### Purchasing Analysis (Gender/Male)
other_data = Heroes_of_Pymoli.query('Gender == "Other / Non-Disclosed"')
other_purchase_count = other_data["Purchase ID"].count()
other_Average_Purchase_Price = other_data["Price"].mean()
other_Total_Purchase_Value = other_data["Price"].sum()
other_Total_Purchase_Value = other_data["Price"].sum()
print("Purchase count for males is " + str(other_purchase_count))
print("Average purchase price for males is " + str(other_Average_Purchase_Price))
print("Total purchase valuefor males is " + str(other_Total_Purchase_Value))
other_data.loc[other_data['SN'] == 1, 'Price'].sum()
print("Average Purchase Total for other " + str(other_data["Price"].mean()))

Purchase count for males is 7
Average purchase price for males is 3.3542857142857145
Total purchase valuefor males is 23.48
Average Purchase Total for other 3.3542857142857145


In [25]:
len(Heroes_of_Pymoli_males)/len(Heroes_of_Pymoli_all)*100

85.02415458937197

In [35]:
#Percentage and Count of Female Players

In [111]:
### Purchasing Analysis (Age)
### Age Demographics

* The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.)
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Average Purchase Total per Person by Age Group

SyntaxError: invalid syntax (<ipython-input-111-2d329d9e986d>, line 4)

In [27]:
bins = pd.cut(Heroes_of_Pymoli['Age'], [4, 8, 12, 16, 20, 24, 28, 32, 36, 40, 44, 48])

new_age_data = Heroes_of_Pymoli.groupby(bins)['Price'].agg(['count', 'mean', 'sum'])
new_age_data.head(11)

count      mean     sum
Age                              
(4, 8]       10  3.749000   37.49
(8, 12]      15  3.221333   48.32
(12, 16]     44  3.002955  132.13
(16, 20]     94  3.183617  299.26
(20, 24]    126  2.986349  376.28
(24, 28]     51  2.772549  141.40
(28, 32]     31  3.075161   95.33
(32, 36]     22  2.795455   61.50
(36, 40]     14  3.581429   50.14
(40, 44]      6  3.305000   19.83
(44, 48]      1  1.700000    1.70

In [28]:
print("Average Purchase Total per Person by Age Group " + str(new_age_data["sum"].mean()))

Average Purchase Total per Person by Age Group 114.85272727272726


In [ ]:
* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value


In [30]:
Heroes_of_Pymoli['Total'] = Heroes_of_Pymoli.groupby(['SN', 'Item Name'])['Price'].transform('sum')
Heroes_of_Pymoli.sort_values(by=['Total'], inplace=True, ascending=False)
Heroes_of_Pymoli.head()

Purchase ID              SN  Age Gender  Item ID  \
554          554        Dyally87   22   Male       63   
110          110         Ririp86   25   Male      139   
231          231  Yarithsurgue62   26   Male      139   
246          246      Lirtilsa71   24   Male      139   
493          493   Chanirrasta87   14   Male      139   

                   Item Name  Price  Total  
554           Stormfury Mace   4.99   4.99  
110  Mercy, Katana of Dismay   4.94   4.94  
231  Mercy, Katana of Dismay   4.94   4.94  
246  Mercy, Katana of Dismay   4.94   4.94  
493  Mercy, Katana of Dismay   4.94   4.94

In [31]:
SN_count= Heroes_of_Pymoli.groupby("SN")
Heroes_of_Pymoli['Purchase Count']= Heroes_of_Pymoli.groupby(['SN'])['SN'].transform('count')
Heroes_of_Pymoli['Average Purchase Price']= Heroes_of_Pymoli.groupby(['SN'])['Price'].transform('mean')
Heroes_of_Pymoli['Total Purchase Value']= Heroes_of_Pymoli.groupby(['SN'])['Price'].transform('sum')

In [33]:
spender_data = Heroes_of_Pymoli.groupby('SN').sum()
spender_data.sort_values(by=['Total'], inplace=True, ascending=False)
new_spenders_data = spender_data.head(5)
new_spenders_data.drop(['Age', 'Purchase ID', 'Item ID', 'Price', 'Total'], axis=1, inplace=True)
new_spenders_data.head()

C:\Users\hanys\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Purchase Count  Average Purchase Price  Total Purchase Value
SN                                                                          
Dyally87                     1                    4.99                  4.99
Lirtilsa71                   1                    4.94                  4.94
Chanirrasta87                1                    4.94                  4.94
Ririp86                      1                    4.94                  4.94
Yarithsurgue62               1                    4.94                  4.94

In [ ]:
* Identify the 5 most popular items by purchase count, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value


In [415]:
#####         new_Item_ID.head()

In [34]:
new_Item_ID['Purchase Count']= new_Item_ID.groupby(['Item Name'])['Item Name'].transform('count')
new_Item_ID['Total Purchase Value']= new_Item_ID.groupby(['Item Name'])['Price'].transform('sum')

In [35]:
spender_data = Heroes_of_Pymoli.groupby('Item Name').sum()
spender_data.sort_values(by=['Purchase Count'], inplace=True, ascending=False)
new_spenders_data = spender_data.head(5)
new_spenders_data.drop(['Age', 'Average Purchase Price', 'Purchase ID'], axis=1, inplace=True)
new_spenders_data.head()

C:\Users\hanys\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Item ID  Price  Total  Purchase Count  \
Item Name                                                                   
Final Critic                            644  32.09  32.09               7   
Fiery Glass Crusader                    870  27.48  27.48               6   
Fate, Vengeance of Eternal Justice      774   9.24   9.24               6   
Oathbreaker, Spellblade of Trials       954  18.48  18.48               6   
Singed Scalpel                          515  21.75  21.75               5   

                                    Total Purchase Value  
Item Name                                                 
Final Critic                                       32.09  
Fiery Glass Crusader                               27.48  
Fate, Vengeance of Eternal Justice                  9.24  
Oathbreaker, Spellblade of Trials                  18.48  
Singed Scalpel                                     21.75

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [36]:
new_Item_ID['Purchase Count']= new_Item_ID.groupby(['Item Name'])['Item Name'].transform('count')
new_Item_ID['Total Purchase Value']= new_Item_ID.groupby(['Item Name'])['Price'].transform('sum')

Purchase ID  SN  Age  Gender  Item Name  Price  Purchase Count  \
Item ID                                                                   
0                  4   4    4       4          4      4              42   
1                  4   4    4       4          4      4              42   
2                  6   6    6       6          6      6              17   
3                  6   6    6       6          6      6              17   
4                  5   5    5       5          5      5              36   
...              ...  ..  ...     ...        ...    ...             ...   
178               12  12   12      12         12     12               1   
179                6   6    6       6          6      6              17   
181                5   5    5       5          5      5              36   
182                3   3    3       3          3      3              30   
183                3   3    3       3          3      3              30   

         Total Purchase Value  
Item ID                        
0                         168  
1                         168  
2                         102  
3                         102  
4                         180  
...                       ...  
178                        12  
179                       102  
181                       180  
182                        90  
183                        90  

[179 rows x 8 columns]

In [62]:
Purchase_Value = Heroes_of_Pymoli.groupby('Item Name').sum()
Purchase_Value.sort_values(by=['Total Purchase Value'], inplace=True, ascending=False)
Total_Purchase_Value = spender_data.head(5)
Total_Purchase_Value.drop(['Age', 'Average Purchase Price', 'Purchase ID'], axis=1, inplace=True)
Total_Purchase_Value.head()

C:\Users\hanys\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Item ID  Price  Total  Purchase Count  \
Item Name                                                                   
Final Critic                            644  32.09  32.09               7   
Fiery Glass Crusader                    870  27.48  27.48               6   
Fate, Vengeance of Eternal Justice      774   9.24   9.24               6   
Oathbreaker, Spellblade of Trials       954  18.48  18.48               6   
Singed Scalpel                          515  21.75  21.75               5   

                                    Total Purchase Value  
Item Name                                                 
Final Critic                                       32.09  
Fiery Glass Crusader                               27.48  
Fate, Vengeance of Eternal Justice                  9.24  
Oathbreaker, Spellblade of Trials                  18.48  
Singed Scalpel                                     21.75

In [72]:
itemName = Total_Purchase_Value.head(5)['Item ID']
itemPurchas = Total_Purchase_Value.head(5)['Purchase Count']
for i in range(0,5):
    Total_Purchase_Value['Item ID'][i]=int(itemName[i]/itemPurchas[i])

C:\Users\hanys\anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.
C:\Users\hanys\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3331: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [73]:
Total_Purchase_Value

Item ID  Price  Total  Purchase Count  \
Item Name                                                                   
Final Critic                             92  32.09  32.09               7   
Fiery Glass Crusader                    145  27.48  27.48               6   
Fate, Vengeance of Eternal Justice      129   9.24   9.24               6   
Oathbreaker, Spellblade of Trials       159  18.48  18.48               6   
Singed Scalpel                          103  21.75  21.75               5   

                                    Total Purchase Value  
Item Name                                                 
Final Critic                                       32.09  
Fiery Glass Crusader                               27.48  
Fate, Vengeance of Eternal Justice                  9.24  
Oathbreaker, Spellblade of Trials                  18.48  
Singed Scalpel                                     21.75

# Findings:
<ul>
    <li>
   Total revenue is 2379.77$
   </li>
   
   <li>
   Final Critic game is the most popular and profiable as well
   </li>
   
   <li>
   Most profitable average age is between 20 to 24 with revenue of 799.74$
        </li>
    </ul>